api/envoy/api/v2/auth/cert.proto

```diff
@@ -279,6 +279,12 @@ message UpstreamTlsContext {
   //
   //   TLS renegotiation is considered insecure and shouldn't be used unless absolutely necessary.
   bool allow_renegotiation = 3;
+
+  // Maximum number of session keys (Pre-Shared Keys for TLSv1.3+, Session IDs and Session Tickets
+  // for TLSv1.2 and older) to store for the purpose of session resumption.
+  //
+  // Defaults to 1, setting this to 0 disables session resumption.
+  google.protobuf.UInt32Value max_session_keys = 4;
```

Comment:
Q: Should this feature be disabled or enabled by default?

Reply:
There is a case of a ["privacy leak"](https://arxiv.org/pdf/1810.07304.pdf) (passive observer being able to correlate connections from the same user by looking at the session that's being resumed, which is sent unencrypted in `ClientHello`) in TLS versions older than 1.3, but that's mostly a threat to end-users (so also a single-user client-side proxy) and not middle/edge proxies and/or service mesh, so I don't think that it justifies having this off by default in Envoy.

Note: TLSv1.3 sends single-use sessions, so the default of 1 is probably too small. Perhaps we could make it vary by default, i.e. 1 if `tls_maximum_protocol_version` is smaller than TLSv1.3, and 4(?) otherwise.

Reply:
I think the common case is service mesh and middle/edge proxies, so we should optimize for that. It's probably not great to be optimizing for TLS 1.3 quiet yet, I don't think this is universal by far.

Reply:
Sorry, I somehow missed this comment earlier.

What I meant regarding TLS v1.3 is basically:
```
if upstream_tls_context.tls_params.tls_maximum_protocol_version == TLSv1_3:
   max_session_keys = 4;
else
   max_session_keys = 1;
```

But I'm fine leaving the default at `1` for the time being, and we can revisit it later, when we enable TLSv1.3 by default.